hail/python/hailtop/batch/docs/cookbook/clumping.rst

```diff
@@ -0,0 +1,293 @@
+and a list containing all of the individual clumping results files. We use the `'ubuntu:18.04'`
+Docker image for this job. The return value is the new :class:`.Job` created.
+
+.. code-block:: python
+
+    def merge(batch, results):
+        """
+        Merge clumped results files together
+        """
+        merger = batch.new_job(name='merge-results')
+        merger.image('ubuntu:18.04')
+        if results:
+            merger.command(f'''
+    head -n 1 {results[0]} > {merger.ofile}
+    for result in {" ".join(results)}
+    do
+        tail -n +2 "$result" >> {merger.ofile}
+    done
+    sed -i '/^$/d' {merger.ofile}
+    ''')
+        return merger
+
+~~~~~~~~~~~~
+Control Code
+~~~~~~~~~~~~
+
+The last thing we want to do is use the functions we wrote above to create new jobs
+on a :class:`.Batch` which can be executed with the :class:`.ServiceBackend`.
+
+First, we create a :class:`.Batch` specifying the backend is the :class:`.ServiceBackend`
+and give it the name 'clumping'.
+
+.. code-block:: python
+
+    backend = hb.ServiceBackend()
+    batch = hb.Batch(backend=backend, name='clumping')
+
+
+Next, we create :class:`.InputResourceFile` objects for the VCF file and
+phenotypes file using the :meth:`.Batch.read_input` method. These
+are the inputs to the entire Batch and are not outputs of a :class:`.Job`.
+We have to specify the extension of the VCF file to Batch as Hail uses
+the file extension in order to read the file. The default is any :class:`.ResourceFile`
+in Batch has no extension.
+
+.. code-block:: python
+
+    vcf = batch.read_input('gs://hail-tutorial/1kg.vcf.bgz')
+    vcf.add_extension('.vcf.bgz')
+
+    phenotypes = batch.read_input('gs://hail-tutorial/1kg_annotations.txt')
+
+We use the `run_gwas` function defined above to create a new job on the batch to
+perform a GWAS that outputs a binary PLINK file and association results:
+
+.. code-block:: python
+
+    gwas = run_gwas(batch, vcf, phenotypes)
+
+We call the `clump` function once per chromosome and aggregate a list of the
+clumping results files passing the outputs from the `gwas` job defined above
+as inputs to the `clump` function:
+
+.. code-block:: python
+
+    results = []
+    for chr in range(1, 23):
+        c = clump(batch, gwas.ofile, gwas.ofile.assoc, chr)
+        results.append(c.clumped)
+
+Finally, we use the `merge` function to concatenate the results into a single file
+and then write this output to a permanent location using :meth:`.Batch.write_output`.
+The inputs to the `merge` function are the clumped output files from each of the `clump`
+jobs.
+
+.. code-block:: python
+
+    m = merge(batch, results)
+    batch.write_output(m.ofile, 'gs://<MY_BUCKET>/batch-clumping/1kg-caffeine-consumption.clumped')
+
+The last thing we do is submit the Batch to the service and then close the Backend:
+
+.. code-block:: python
+
+    batch.run(open=True, wait=False)  # doctest: +SKIP
+    backend.close()
```

Comment:
we should include the Dockerfile and complete script at the bottom for easy copy-pasting.